src/core/net/srp_advertising_proxy.cpp

```diff
@@ -0,0 +1,1393 @@
+
+    bool didUpdate = false;
+
+    VerifyOrExit(aHost.mAdvIdRange.Contains(aRequestId));
+
+    if (aHost.mAdvId == aRequestId)
+    {
+        aHost.mAdvId        = kInvalidRequestId;
+        aHost.mIsRegistered = (aError == kErrorNone);
+        didUpdate           = true;
+    }
+
+    if (aHost.mKeyAdvId == aRequestId)
+    {
+        aHost.mKeyAdvId        = kInvalidRequestId;
+        aHost.mIsKeyRegistered = true;
+        didUpdate              = true;
+    }
+
+    for (Service &service : aHost.mServices)
+    {
+        if (service.mAdvId == aRequestId)
+        {
+            service.mAdvId        = kInvalidRequestId;
+            service.mIsRegistered = (aError == kErrorNone);
+            didUpdate             = true;
+        }
+
+        if (service.mKeyAdvId == aRequestId)
+        {
+            service.mKeyAdvId        = kInvalidRequestId;
+            service.mIsKeyRegistered = true;
+            didUpdate                = true;
+        }
+    }
+
+    UpdateAdvIdRangeOn(aHost);
+
+exit:
+    return didUpdate;
+}
+
+void AdvertisingProxy::HandleTimer(void)
+{
+    TimeMilli           now      = TimerMilli::GetNow();
+    TimeMilli           nextTime = now.GetDistantFuture();
+    OwningList<AdvInfo> expiredList;
+
+    VerifyOrExit(mState == kStateRunning);
+
+    mAdvInfoList.RemoveAllMatching(AdvInfo::ExpirationChecker(now), expiredList);
+
+    for (AdvInfo &adv : mAdvInfoList)
+    {
+        nextTime = Min(adv.mExpireTime, nextTime);
+    }
+
+    if (nextTime != now.GetDistantFuture())
+    {
+        mTimer.FireAtIfEarlier(nextTime);
+    }
+
+    for (AdvInfo &adv : expiredList)
+    {
+        adv.mError       = kErrorResponseTimeout;
+        adv.mBlockingAdv = nullptr;
+        adv.mHost.mAdvIdRange.Clear();
+        SignalAdvCompleted(adv);
+    }
+
+exit:
+    return;
+}
+
+void AdvertisingProxy::HandleTasklet(void)
+{
+    VerifyOrExit(mState == kStateRunning);
+
+    while (true)
+    {
+        OwningList<AdvInfo> completedList;
+
+        mAdvInfoList.RemoveAllMatching(AdvInfo::CompletionChecker(), completedList);
+
+        VerifyOrExit(!completedList.IsEmpty());
+
+        // `RemoveAllMatching()` reverses the order of removed entries
+        // from `mAdvInfoList` (which itself keeps the later requests
+        // towards the head of the list). This means that the
+        // `completedList` will be sorted from earliest request to
+        // latest and this is the order that we want to notify
+        // `Srp::Server`.
+
+        for (AdvInfo &adv : completedList)
+        {
+            SignalAdvCompleted(adv);
+        }
+
+        completedList.Clear();
+    }
+
+exit:
+    return;
+}
+
+void AdvertisingProxy::SignalAdvCompleted(AdvInfo &aAdvInfo)
+{
+    // Check if any outstanding advertisements in the list
+    // is blocked by `aAdvInfo` and unblock.
+
+    for (AdvInfo &adv : mAdvInfoList)
+    {
+        if (adv.mBlockingAdv == &aAdvInfo)
```

Comment:
IIUC the `mBlockingAdv` only blocks marking the completion of the advertisement, while the mDNS operations are not blocked. I wonder if it also needs to block the mDNS operations?

Reply:
We don't need to block mDSN (`otPlatDnssd`) operations. The `CompareAndUpdateHostAndServices()` method already takes care of this. This method determines if the new entries (in a new SRP update) replace, change, or remove any existing entries (whether already registered or still being registered) and handles all cases.

The `mBlockingAdv` is used to ensure that "SRP Update" messages that are related (i.e., they change the same host or services) are committed in the order that they are received. When we set `mBlockingAdv` on an `AdvInfo`, we make sure it won't be committed even if its advertisement finishes early and before the earlier related one is committed.

Example situation:
- We receive an SRP Update from a client registering two services, A and B.
- While waiting for the this to advertised, we receive a new SRP Update that removes service B.
- Unregistering the service on `DnssdPlat` is often quick (no need probe), so the new advertisement for removing B will complete quickly.
- We need to make sure that these two SRP updates are committed in order.
- So we mark the second one to be blocked by the first one.
- Otherwise, we could end up with a situation where a removed service (B in this example) is incorrectly committed on the SRP server.

This situation of getting back-to-back messages that modify the same host/services will become a more common scenario with SRP Replication (when a new SRPL partner syncs, it gets all messages from the peer from the same client).

Reply:
BTW, the new unit test `test_srp_adv_proxy` added in this PR, covers a bunch of these scenarios in `TestSrpAdvProxyReplacedEntries()`.